extensions/jdbc/jdbc-oracle/runtime/src/main/java/io/quarkus/jdbc/oracle/runtime/RollbackOnConnectionClosePoolInterceptor.java

```diff
@@ -0,0 +1,62 @@
+package io.quarkus.jdbc.oracle.runtime;
+
+import java.sql.Connection;
+
+import javax.sql.XAConnection;
+
+import org.jboss.logging.Logger;
+
+import io.agroal.api.AgroalPoolInterceptor;
+import io.agroal.pool.wrapper.ConnectionWrapper;
+import oracle.jdbc.OracleConnection;
+
+/**
+ * Oracle has the weird behavior that it performs an implicit commit on normal connection closure (even with auto-commit
+ * disabled),
+ * which happens on application shutdown. To prevent whatever intermittent state we have during shutdown from being committed,
+ * we add an explicit rollback to each connection closure. If the connection has already received a COMMIT, the rollback will
+ * not work, which is fine.
+ * <p>
+ * The code unwraps the {@link Connection} so that we perform the rollback directly on the underlying database connection,
+ * and not on e.g. Agroal's {@link ConnectionWrapper} which can prevent the rollback from actually being executed due to some
+ * safeguards.
+ *
+ * @see https://github.com/quarkusio/quarkus/issues/36265
+ */
+public class RollbackOnConnectionClosePoolInterceptor implements AgroalPoolInterceptor {
+
+    private static final Logger LOG = Logger.getLogger(RollbackOnConnectionClosePoolInterceptor.class);
+
+    private final boolean noAutomaticRollback;
+
+    public RollbackOnConnectionClosePoolInterceptor() {
+        // if you have to use this system property, make sure you open an issue in the Quarkus tracker to explain
+        // why as we might need to adjust things
+        noAutomaticRollback = Boolean.getBoolean("quarkus-oracle-no-automatic-rollback-on-connection-close");
+    }
+
+    @Override
+    public void onConnectionDestroy(Connection connection) {
+        if (noAutomaticRollback) {
+            return;
+        }
+
+        if (connection instanceof ConnectionWrapper connectionWrapper) {
+            if (connectionWrapper.getHandler().getXaResource() != null) {
+                return;
+            }
+        }
+
+        try {
+            if (connection.unwrap(Connection.class) instanceof OracleConnection oracleConnection) {
+                if (connection.isClosed() || connection.getAutoCommit()) {
+                    return;
+                }
+
```

Comment:
I see an `instanceof`, my first thought is to get rid of it :]

In this case... move it off the "normal" path?

```suggestion
            if (connection.isClosed() || connection.getAutoCommit()) {
                return;
            }

            if (connection.unwrap(Connection.class) instanceof OracleConnection oracleConnection) {
```

Reply:
Yeah so.. I did it that way because the logic of the change is to only do something if it's an Oracle connection. Thus why it's implemented this way.

Now I suppose it's simple enough that it doesn't hurt to do it your way.

Reply:
So actually, I went back to my initial coding, mostly because `isClosed()` and `getAutoCommit()` can potentially throw `SQLException` so they might not simple getters and in this case we should avoid doing the work if it's not an Oracle connection.